src/main/java/tdl/client/queue/QueueBasedImplementationRunner.java

```diff
@@ -19,11 +21,89 @@
     private final Audit audit;
     private final ProcessingRules deployProcessingRules;
     private final ImplementationRunnerConfig config;
+    private long consumedMessages;
+    private long receivesMessages;
+    private List<Response> responses;
+    private List<Request> requests;
+    private RemoteBroker remoteBroker;
+    private MessageProcessingThread messageProcessingThread;
 
     private QueueBasedImplementationRunner(ImplementationRunnerConfig config, ProcessingRules deployProcessingRules) {
+        logToConsole("        QueueBasedImplementationRunner creation [start]");
         this.config = config;
         this.deployProcessingRules = deployProcessingRules;
         audit = new Audit(config.getAuditStream());
+
+        responses = new ArrayList<>();
+        requests = new ArrayList<>();
+
+        logToConsole("        QueueBasedImplementationRunner creation [end]");
+    }
+
+    public long getConsumedMessagesCount() {
+        return consumedMessages;
+    }
+
+    public long getReceivesMessagesCount() {
+        return receivesMessages;
+    }
+
+    public List<String> getReceivedMessages() {
+        List<String> results = new ArrayList<>();
+
+        for (Response response : responses) {
+
+            String result;
+            if ((response.getResult() == null) ||
+                    isNumeric(response.getResult())) {
+                result = String.format(
+                        "{\"result\":%s,\"error\":null,\"id\":\"%s\"}",
+                        response.getResult(), response.getId());
+            } else {
+                result = String.format(
+                        "{\"result\":\"%s\",\"error\":null,\"id\":\"%s\"}",
+                        response.getResult(), response.getId());
+            }
+
+            results.add(result);
+        }
+        responses.clear();
+
+        return results;
+    }
+
+    public void start() {
+        try {
+            requests.clear();
+            remoteBroker = new RemoteBroker(config.getHostname(), config.getPort(), config.getUniqueId(), config.getRequestTimeoutMillis());
+            messageProcessingThread = new MessageProcessingThread(remoteBroker, audit, requests);
```

Comment:
Should not use a MessageProcessingThread, everything happens on the main thread

Reply:
Wouldn't this make the server blocking and then it won't return? So the tests won't be able to assert on the results.

Unless you want the server to read all the messages in the queue and only finish when the queue is empty and not read anymore? But the server is always listening and blocking. Sort of contradicts the use case.

Reply:
And how do we maintain this while keeping it non-blocking, moving functionality to the main thread can make it blocking.

In the test cases, the scenarios are single events based or serial events based. But in the real world we can we have a sequence of messages with very small gaps sent one after the other. How will the server wake up each time such messages arrive in the queue? 

Should work both for production and when test invokes the server.